index.bs

```diff
@@ -2015,14 +1994,113 @@ Issue: This needs to be generalised to work with realms too.
 <div algorithm="remote end steps for session.unsubscribe">
 The [=remote end steps=] with |session| and |command parameters| are:
 
-1. Let the |list of event names| be the value of the <code>events</code> field of
-   |command parameters|.
+1. If |command parameters| matches the <code>session.UnsubscribeByAttributesRequest</code> production:
+
+   1. Let |event names| be a [=/set=] of |command parameters|["<code>events</code>"].
+
+   1. Let |top-level traversable context ids| be a [=/set=].
+
+   1. Let |input context ids| be a [=/set=] of |command parameters|[<code>contexts</code>].
+
+   1. If |input context ids| is not empty:
+
+      1. Let |navigables| be [=get navigables by ids=] given |input context ids|.
+
+      1. If [=list/size=] of |navigables| does not equal [=list/size=] of |input context ids|:
+
+         1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Set |top-level traversables| be [=get top-level traversables=] given |navigables|.
+
+      1. For each |navigable| in |top-level traversables|:
+
+         1. [=set/Append=] |navigable|'s [=navigable id=] to |top-level traversable context ids|.
+
+   1. Let |new subscriptions| to be a [=/list=].
+
+   1. For each |subscription| of |session|'s [=subscriptions=]:
+
+      1. If [=set/intersection=] of |subscription|'s [=subscription/event names=] and |event names| is an empty [=/set=]:
+
+         1. [=list/append=] |subscription| to |new subscriptions|.
 
-1. Let the |list of contexts| be the value of the <code>contexts</code>
-   field of |command parameters| if it is present or null if it isn't.
+         1. Continue.
 
-1. [=Try=] to [=update the event map=] with |session|,
-   |list of event names|, |list of contexts| and enabled false.
+      1. If |top-level traversable context ids| is an empty [=/set=]:
+
+        1. [=list/Remove=] all items [=list/contains|contained=] in |event names| from |subscription|'s [=subscription/event names=].
+
+        1. If |subscription|'s [=subscription/event names=] is not empty:
```

Comment:
Otherwise shouldn't we add the subscription id to `internally removed subscription ids`? Otherwise you can have code like:
```
// In the harness
subscribe(events=[A])
// In a test
subId = subscribe(events=[A])
// In the harness, for some reason
unsubscribe(events=[A])
// In the test teardown
unsubscribe(subscription=subId)
```
and the second unsubscribe will fail, even though you passed in a totally valid handle. I think it should be an invariant that you only get an error if you use a subscription id to subscribe more than once, even if some subsequent action made the subscription a no-op (it's problematic either way to mix unsubscribing by ids and events, but I think this is less bad because at least a test fixture can gather a list of subscriptions it created and know that passing them to unsubscribe won't break anything. Otherwise I think we should at least return a set of subscription ids that were invalidated, but that's a weird model).

Reply:
My idea was that we would still want to error if unsubscribe happened via commands indicating some client bugs/redundant calls but only suppress errors if we removed a subscription due to navigables being disposed and other internal reasons.

So in this scenario, I would indeed expect an error in the last unsubscribe as the subscriptions has already been removed by the client which indicates redundant client calls. I think it could be viewed as sending `unsubscribe(subscription=subId)` twice.

Reply:
Anyway, I think either way works and tracking known subscription ids makes the spec easier, so I adopted your proposal into the text.